# One Hot encoding - Variable with many Categories

In [2]:
import numpy as np
import pandas as pd



In [5]:
# Let' see the data with few categories columns to perform one-hot encoding
df = pd.read_csv("car_data.csv", usecols = ['X1','X2','X3','X4','X5','X6'])

df.head()

X1  X2 X3 X4 X5 X6
0  v  at  a  d  u  j
1  t  av  e  d  y  l
2  w   n  c  d  x  j
3  t   n  f  d  x  l
4  v   n  f  d  h  d

In [8]:
# Now firstly i will check the uniques values {labels} of each columns
for col in df.columns:
    print(col, ':', len(df[col].unique()), 'labels')


X1 : 28 labels
X2 : 45 labels
X3 : 8 labels
X4 : 5 labels
X5 : 30 labels
X6 : 13 labels


In [10]:
# let's check while creating dummy variable through one-hot encoding in simple way
# Also will check the shape
df2 = pd.get_dummies(df, drop_first= True)
df2.head()

X1_aa  X1_ab  X1_b  X1_c  X1_d  X1_e  X1_f  X1_g  X1_h  X1_i  ...  X6_c  \
0      0      0     0     0     0     0     0     0     0     0  ...     0   
1      0      0     0     0     0     0     0     0     0     0  ...     0   
2      0      0     0     0     0     0     0     0     0     0  ...     0   
3      0      0     0     0     0     0     0     0     0     0  ...     0   
4      0      0     0     0     0     0     0     0     0     0  ...     0   

   X6_d  X6_e  X6_f  X6_g  X6_h  X6_i  X6_j  X6_k  X6_l  
0     0     0     0     0     0     0     1     0     0  
1     0     0     0     0     0     0     0     0     1  
2     0     0     0     0     0     0     1     0     0  
3     0     0     0     0     0     0     0     0     1  
4     1     0     0     0     0     0     0     0     0  

[5 rows x 117 columns]

In [11]:
df2.shape

(4212, 117)

Well, as you can. While done with simple One-hot Encoding, now we have total 117 columns, which is too big. We were started with only 05 columns and now got more 112 columns, which is not so good.

As we all knows very well, as much the columns size will get increased that may cause **cursive dimensionality**, which would not good to predict great score or accurary score.

**The curse of dimensionality refers to various phenomena that arise when analyzing and organizing data in high-dimensional spaces that do not occur in low-dimensional settings**

Well, let's move ahead and see the further phenomena

In [13]:
# let's find the top 10 most frequent categories for the variable X2.
df.X2.value_counts().sort_values(ascending=False).head(20)

as    1659
ae     496
ai     415
m      367
ak     265
r      153
n      137
s       94
f       87
e       81
aq      63
ay      54
a       47
t       29
i       25
k       25
b       21
ao      20
ag      19
z       19
Name: X2, dtype: int64

In [14]:
# Let's take the top 10 and convert that into list with top varibles
top_10 = [x for x in df.X2.value_counts().sort_values(ascending=False).head(20).index]
top_10

['as',
 'ae',
 'ai',
 'm',
 'ak',
 'r',
 'n',
 's',
 'f',
 'e',
 'aq',
 'ay',
 'a',
 't',
 'i',
 'k',
 'b',
 'ao',
 'ag',
 'z']